### ISSUE TITLE
Simple stuff like margins

### ISSUE THREAD

**Issue by clintolsen** [simoncozens/sile#460]
I searched through the SILE book, and aside from references to it in Lua, I could not find basic commands to control left/right/top/bottom margins. If I missed it, please send me a cluepon. Thanks...

**Comment by alerque** [comment]
What you're looking for isn't called margins at all. You want to setup a frameset that describes your page layout. This can be a single frame or multiple frames to include headers, footers, columns, etc. The text generally gets placed inside the frameset right up to the edge (with the exception of various skip options), so your overall page layout (what you called left/right/top/bottom margins) is actually configured by setting the edges of your content frame relative to the page size.

You're quite right this could use better documentation so lets keep the issue open to make a more explicit note about this so people searching for these terms will find the right settings.

**Comment by StefanSchroeder** [comment]
Well, 

[https://en.wikipedia.org/wiki/Margin_(typography)](https://en.wikipedia.org/wiki/Margin_(typography))

seems to indicate that it is indeed called margins.

**Comment by alerque** [comment]
@clintolsen Please review the copy I added to the manual  in #462 and see if those explanations address your concern. Would searching for "margins" and finding those blurbs have pointed you in the right direction of what tool to reach for?

**Comment by simoncozens** [comment]
This is all good, but it might be a bit more user friendly to have a package similar to TeX's "[geometry](http://texdoc.net/texmf-dist/doc/latex/geometry/geometry.pdf)" that makes it easy for you to define a page layout without having to do too much Lua. It doesn't really fit with the "document content in XML, layout in Lua" philosophy but, hey, people will want it and it'll end up happening anyway so we may as well do it.

**Comment by alerque** [comment]
I agree. There are actually a couple of setup things like this that could be bridged with a package. On the other hand this particular instance –at least in my head– also makes some sense as a class option. The class already takes a paper size option, why shouldn't it take some other parameters as well so that values like default page margins aren't hard coded into the class? Each class might have different options depending on the purpose.

Another example is crop marks. SILE currently has this as a package, but I ran into some series downsides to having this configured after the initial class was setup, so in CaSILE I actually rewrote it to be a class option. I realize you can hook into and _redo_ things the class setup, but this has some drawbacks. One of them being other packages that need to know what mode to run it might get initialized incorrectly. Without the equivalent of TeX's preamble that runs before the start of the document some things are a bit awkward to do in packages.

**Comment by clintolsen** [comment]
I think the text you added is fine. I agree with others that Sile is eventually going to have to support parameters of some kind in order to control some aspects of page layout. I think asking non-techie users to write Lua to do that is a bit of an ask.

**Comment by hurzl** [comment]
More than 7 years later, I haven't found this hint for beginners because I am not at all interested in a "MS Word" comparison.

**Comment by Omikhleia** [comment]
>  And now I still haven't found out how to define these frames ...

You might not have searched hard enough ;) 

- SILE manual §10.2.4, on how to define your (possibly book-derived) class, has a section on adapting the frame set.
- Discussions #1601 and #1919 also tackled some aspects of it -- the latter, notably, shows an example with page masters.
- Moreover, allow me to note that [resilient.sile](https://github.com/Omikhleia/resilient.sile) proposes yet another way (page layouts, pre-defined and custom)

**Comment by hurzl** [comment]
>     * SILE manual §10.2.4, on how to define your own (possibly book-derived) class, has a section on adapting the frame set.

I _have_ found this, but it is the end of the book, and I don't want to design a class and I have no idea how to apply these settings.
I just want to adjust margins. The most basic thing next to paper format.

**Comment by Omikhleia** [comment]
>  I don't want to design a class (...) I just want to adjust margins. The most basic thing next to paper format.

Sure. Then why not installing and using  the resilient 3rd party collection as I suggested, and benefit from the very simple user-friendly layouts? For a custom (head margin, inner margin, footer margin, outer margin):

```
\begin[class=resilient.book, layout=geometry 1in 1in 1in 1in, papersize=a5]{document}

My document ...

\end{document}
```

That's the most user-friendly way I can think to :rofl: and there's a lot more advanced settings under the hood.

**Comment by hurzl** [comment]
So you demand either importing an extra package or programming an extra class with an extra file for every single document? I can't think of any case where I would keep the page margins as in the default settings for anything I would want to typeset.

**Comment by Omikhleia** [comment]
> So you demand either importing an extra package (...)

I'm afraid I'm not sure I fully understand your concern and where this discussion in a long-closed issue is heading to.

If it may helps, let me rephrase my suggestions with a comparison, _mutatis mutandis_.[^1]
Would you be using the TeX engine,

 1. would you set the margins via low level TeX macro commands at the top of your document? That's what my second solution mentioning two discussions equates too (= with either `\pagetemplate` or page `masters`).
 2. would you rather install LaTeX on top of TeX, and design you own style/class file in LaTeX internal macro commands? That's more or less what the first solution is about (= small inherited class in Lua in our case, and the few lines of codes are also in one the above-mentioned discussions)
 3. or would you use LateX but additionally install the "geometry" package from CTAN, for a higher level approach? My suggestion using the "resilient" collection (to use its geometry layout) is fairly similar in aim.

I am not mistaking, you went through building SILE from sources (#2107), so installing a few add-ons is within your reach.[^2] It boils down to:

```
# Check the Lua version SILE uses, depending on how you built it...
sile -q -e 'print(SILE.lua_version); os.exit()' 
# Install a 3rd-party add on, where X.Y is what the previous command returned
luarocks install resilient.sile --lua-version X.Y
```

I hope this clarifies my previous answer a bit.
Cheers.[^3]

[^1]: I am certainly not really claiming here  that SILE=TeX and resilient=LaTeX, that's obviously an over simplification for the sake of comparing strategies :eyes: 

[^2]: As stated, the comparison is limited. (People usually get their TeX engines from the TeXlive distribution, which also comes with a lot of CTAN-packages _already_ bundled with it. Well, the TeX ecosystem, of course, has several decades of involvement from various contributors. Hey, maybe one day we'll have a SILElive distribution too, who knows?)

[^3]: As a final comment, "page margins" as defined here are not the simplest things in the world actually. Such solutions would not work with highly complex layouts where the text block is not "just" a square in the page.

**Comment by alerque** [comment]
> So you demand either importing an extra package

You can change the margins *inside* the frame with a single settings call. Or I should say *relative* to the frame, because a negative margin will actually hang things outside of the frame.

You can also create a new frame to your specifications with a single command, but the original frameset will still be around. The cleanest thing to do for use cases is define your own frameset, probably by replacing the default one for the class. You can do this inline in a document if you want, but creating a new class for it makes it easier to have the frameset all correct before your first page gets initialized.

Also note that you can create a package inside a document just for use by that document, there is no requirement that it be loaded from external sources. Any `\lua{}` code can return a package.

Creating a class is not a huge chore either, you don't have to "design a new class", you just take the closest starting point (e.g. *plain* or *book*) and change or addon to it as you please.

**Comment by hurzl** [comment]
\pagetemplate or page masters equates to plain TeX's  \hsize,\vsize,\hoffset,\voffset? How?

```
resilient.sile 2.4.0-1 is already installed in /usr/local

! Unable to find module 'resilient.book'
```

**Comment by Omikhleia** [comment]
Re: "Unable to find" --> Running sile with `-t` normally shows a stack trace with explored paths. On a custom-built SILE, it's hard to tell sometimes whether paths and env. variables are properly set for luarocks-installed modules. On my system it's in `/usr/share/lua/5.1/sile/classes/resilient/book.lua`  and my stock Archlinux SILE (i.e. `SILE v0.15.5 (LuaJIT 2.1.1723675123) [Rust]`) finds it (likewise for all dependencies).

Re: TeX/SILE comparison --> I did say the comparison had its limits, didn't I? You cannot push it too far either.

N.B. You're still commenting on a closed issue, with one-liners. Am I sensing a disturbance in the force?

**Comment by hurzl** [comment]
```
	no file '/usr/local/share/lua/5.1/sile/resilient/book.lua'
	no file '/usr/local/share/lua/5.1/sile/resilient/book/init.lua'
	no file '/usr/local/lib/lua/5.1/sile/resilient/book.lua'
	no file '/usr/local/lib/lua/5.1/sile/resilient/book/init.lua'
	no file '/usr/local/share/sile/lua_modules/share/lua/5.1/sile/resilient/book.lua'
	no file '/usr/local/share/sile/lua_modules/share/lua/5.1/sile/resilient/book/init.lua'
	no file '/usr/local/share/sile/lua_modules/share/lua/5.1/resilient/book.lua'
	no file '/usr/local/share/sile/lua_modules/share/lua/5.1/resilient/book/init.lua'
	no file '/usr/local/lib/sile/lua-libraries/resilient/book.lua'
	no file '/usr/local/lib/sile/lua-libraries/resilient/book/init.lua'
	no file '/usr/local/lib/sile/resilient/book.lua'
	no file '/usr/local/lib/sile/resilient/book/init.lua'
	no file '/usr/local/lib/sile/lua_modules/share/lua/5.1/sile/resilient/book.lua'
	no file '/usr/local/lib/sile/lua_modules/share/lua/5.1/sile/resilient/book/init.lua'
	no file '/usr/local/lib/sile/lua_modules/share/lua/5.1/resilient/book.lua'
	no file '/usr/local/lib/sile/lua_modules/share/lua/5.1/resilient/book/init.lua'
	no file '/usr/local/share/sile/lua-libraries/resilient/book.lua'
	no file '/usr/local/share/sile/lua-libraries/resilient/book/init.lua'
	no file '/usr/local/share/sile/resilient/book.lua'
	no file '/usr/local/share/sile/resilient/book/init.lua'
	no file '/usr/local/share/lua/5.1/resilient/book.lua'
	no file '/usr/local/share/lua/5.1/resilient/book/init.lua'
	no file '/usr/local/lib/lua/5.1/resilient/book.lua'
	no file '/usr/local/lib/lua/5.1/resilient/book/init.lua'
```

```
$ ls -l /usr/local/share/lua/5.1/sile/classes/resilient/book.lua
-rw-r--r--  1 root wheel 26882 14 Sep. 20:05 /usr/local/share/lua/5.1/sile/classes/resilient/book.lua
```

I guess I give up.

**Comment by Omikhleia** [comment]
> I guess I give up.

It's your call, unfortunately, although you were likely quite close (a path issue with a custom build, or a more intricate dependency issue? Hard to tell with truncated logs, surely somethings is amiss somewhere!). Your user experience would have been useful to help us improve something, surely. After all, SILE 0.15 is quite recent; and my own support for it in _resilient_ (a 3-year long endeavor) is only 2-week old, there might be hiccups using the latest releases... So I'm not blaming your or anyone here. Just remember: the path to open source is full of good wills but deadly pits sometimes ;)

Please feel free to open a dedicated issue, with a clear scope, rather than grunting on an old closed issue, and I'll sure be happy to help if need be. This being said, I'm going to mark some discussions here as out of topic -- Whatever this old closed issue was at its own time, it's unrelated to the present (heh! I wasn't even here 7 years ago, can't even tell which version of SILE was at stake then. _Smile_).
